discovery/file/file_test.go

```diff
@@ -15,158 +15,456 @@ package file
 
 import (
 	"context"
+	"encoding/json"
 	"io"
+	"io/ioutil"
 	"os"
 	"path/filepath"
+	"sort"
+	"sync"
 	"testing"
 	"time"
 
 	"github.com/prometheus/common/model"
 
 	"github.com/prometheus/prometheus/discovery/targetgroup"
+	"github.com/prometheus/prometheus/util/testutil"
 )
 
-const testDir = "fixtures"
+const defaultWait = time.Second
 
-func TestFileSD(t *testing.T) {
-	defer os.Remove(filepath.Join(testDir, "_test_valid.yml"))
-	defer os.Remove(filepath.Join(testDir, "_test_valid.json"))
-	defer os.Remove(filepath.Join(testDir, "_test_invalid_nil.json"))
-	defer os.Remove(filepath.Join(testDir, "_test_invalid_nil.yml"))
-	testFileSD(t, "valid", ".yml", true)
-	testFileSD(t, "valid", ".json", true)
-	testFileSD(t, "invalid_nil", ".json", false)
-	testFileSD(t, "invalid_nil", ".yml", false)
+type testRunner struct {
+	*testing.T
+	dir           string
+	ch            chan []*targetgroup.Group
+	done, stopped chan struct{}
+	cancelSD      context.CancelFunc
+
+	mtx        sync.Mutex
+	tgs        map[string]*targetgroup.Group
+	receivedAt time.Time
 }
 
-func testFileSD(t *testing.T, prefix, ext string, expect bool) {
-	// As interval refreshing is more of a fallback, we only want to test
-	// whether file watches work as expected.
-	var conf SDConfig
-	conf.Files = []string{filepath.Join(testDir, "_*"+ext)}
-	conf.RefreshInterval = model.Duration(1 * time.Hour)
+func newTestRunner(t *testing.T) *testRunner {
+	t.Helper()
 
-	var (
-		fsd         = NewDiscovery(&conf, nil)
-		ch          = make(chan []*targetgroup.Group)
-		ctx, cancel = context.WithCancel(context.Background())
-	)
-	go fsd.Run(ctx, ch)
+	tmpDir, err := ioutil.TempDir("", "prometheus-file-sd")
+	testutil.Ok(t, err)
 
-	select {
-	case <-time.After(25 * time.Millisecond):
-		// Expected.
-	case tgs := <-ch:
-		t.Fatalf("Unexpected target groups in file discovery: %s", tgs)
+	return &testRunner{
+		T:       t,
+		dir:     tmpDir,
+		ch:      make(chan []*targetgroup.Group),
+		done:    make(chan struct{}),
+		stopped: make(chan struct{}),
+		tgs:     make(map[string]*targetgroup.Group),
 	}
+}
+
+// copyFile atomically copies a file to the runner's directory.
```

Comment:
if the it's always the runners directory can't `filepath.Base(tc.dir)` just be a value on `testRunner`?